### ISSUE TITLE
Istio generates invalid JWT authentication filter config when remote JWKS endpoint is unavailable

### ISSUE THREAD

**Issue by haines** [istio/istio#24629]
**Bug description**

When istiod attempted to fetch the JWKS for an issuer specified in a JWT rule, the issuer service responded with a 502.

The istio-proxy pod on our service is now rejecting the LDS config with the error

```
gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected:
Error adding/updating listener(s) virtualInbound:
Proto constraint validation failed (JwtAuthenticationValidationError.Providers[key]: ["embedded message failed validation"]
| caused by JwtProviderValidationError.LocalJwks: ["embedded message failed validation"]
| caused by DataSourceValidationError.InlineString: ["value length must be at least " '\x01' " bytes"])
```

The offending config is

```
    local_jwks {
      inline_string: ""
    }
```

This is particularly problematic because now the proxy container cannot become ready, so restarting the deployment hangs.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[x] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

**Expected behavior**

Istio uses the last good JWKS fetched from that URL. If no good JWKS has been fetched, Istio applies a valid config that results in no JWTs from that issuer being validated. I would say this could be an empty JWKS `{"keys":[]}` except that I seem to remember (in 1.4, at least) these also break the JWT authentication filter - that may have been fixed since though!

**Steps to reproduce the bug**

On a fresh cluster I ran

```console
$ istioctl install

$ kubectl label namespace default istio-injection=enabled

$ kubectl apply -f https://raw.githubusercontent.com/istio/istio/release-1.6/samples/httpbin/httpbin.yaml

$ kubectl apply -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: RequestAuthentication
metadata:
  name: httpbin
spec:
  selector:
    matchLabels:
      app: httpbin
  jwtRules:
  - issuer: https://example.com
EOF

$ istioctl proxy-status
NAME                                 CDS        LDS                            EDS        RDS          PILOT                       VERSION
httpbin-78b9f6c986-gg55q.default     SYNCED     STALE (Never Acknowledged)     SYNCED     SYNCED       istiod-7599c64cb4-lj2qs     1.6.2
...

$ kubectl logs deployment/httpbin istio-proxy
...
2020-06-12T11:00:22.402775Z	warning	envoy config	[external/envoy/source/common/config/grpc_subscription_impl.cc:101] gRPC config for type.googleapis.com/envoy.api.v2.Listener rejected: Error adding/updating listener(s) virtualInbound: Proto constraint validation failed (JwtAuthenticationValidationError.Providers[key]: ["embedded message failed validation"] | caused by JwtProviderValidationError.LocalJwks: ["embedded message failed validation"] | caused by DataSourceValidationError.InlineString: ["value length must be at least " '\x01' " bytes"]): providers {
  key: "origins-0"
  value {
    issuer: "https://example.com"
    local_jwks {
      inline_string: ""
    }
    payload_in_metadata: "https://example.com"
  }
}
rules {
  match {
    prefix: "/"
  }
  requires {
    requires_any {
      requirements {
        provider_name: "origins-0"
      }
      requirements {
        allow_missing {
        }
      }
    }
  }
}

$ kubectl rollout restart deployment/httpbin
deployment.apps/httpbin restarted

$ kubectl get pods
NAME                       READY   STATUS    RESTARTS   AGE
httpbin-6db444f69-45w8t    1/2     Running   0          41s
httpbin-78b9f6c986-gg55q   2/2     Running   0          37m
```

The new pod never becomes ready because the istio-proxy container is not ready.

**Version (include the output of `istioctl version --remote` and `kubectl version` and `helm version` if you used Helm)**

```console
$ istioctl version --remote
client version: 1.6.2
control plane version: 1.6.2
data plane version: 1.6.2 (4 proxies)

$ kubectl version
Client Version: version.Info{Major:"1", Minor:"18", GitVersion:"v1.18.3", GitCommit:"2e7996e3e2712684bc73f0dec0200d64eec7fe40", GitTreeState:"clean", BuildDate:"2020-05-21T14:51:23Z", GoVersion:"go1.14.3", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"16+", GitVersion:"v1.16.6-beta.0", GitCommit:"e7f962ba86f4ce7033828210ca3556393c377bcc", GitTreeState:"clean", BuildDate:"2020-01-15T08:18:29Z", GoVersion:"go1.13.5", Compiler:"gc", Platform:"linux/amd64"}
```

**How was Istio installed?**

```console
$ istioctl install
```

**Environment where bug was observed (cloud vendor, OS, etc)**

Initially observed in EKS, but reproduced easily on Docker for Mac.

**Comment by incfly** [comment]
So the summary is jwt filter config is output correctly as empty "", but the envoy is hanging, which is unexpected. And we should expect envoy come up but jwt filter reject all http request

I thought the expected behavior is what we have now based on our previous discussion. Or this is a regression? @yangminzhu

**Comment by haines** [comment]
I don't think outputting an empty string is valid though, the `local_jwks.inline_string` must have a minimum length of 1 according to the protobuf definition 
- `local_jwks` is defined as a `DataSource` [here](https://github.com/envoyproxy/envoy/blob/9e8b32f3f29213e66dd32a07d5922dca700ea0e2/api/envoy/extensions/filters/http/jwt_authn/v3/config.proto#L119)
- `DataSource.local_string` is required to have `min_bytes: 1` [here](https://github.com/envoyproxy/envoy/blob/9e8b32f3f29213e66dd32a07d5922dca700ea0e2/api/envoy/config/core/v3/base.proto#L320)

I think the correct behaviour would be to set that string to an empty JWKS, `{ "keys": [] }`, assuming Envoy will do the right thing with that (it used to fail with an empty keyset, but that's probably worth fixing in Envoy if it's still the case).

**Comment by yangminzhu** [comment]
> I thought the expected behavior is what we have now based on our previous discussion. Or this is a regression? @yangminzhu

Not a regression, but looks like a bad user experience.

> I think the correct behaviour would be to set that string to an empty JWKS

@haines , thank you for the detail report. I agree it would be a better user experience to let Envoy to reject the user request instead of getting stuck on the xDS update. We do not need to rely on the Envoy implementation but could just generate a different JWT filter config that rejects every request in this case.

**Comment by incfly** [comment]
> We do not need to rely on the Envoy implementation but could just generate a different JWT filter config that rejects every request in this case.

wondering how so?

re better UX: Can someone else argue Istio lets the sidecar become ready but some inbound actually unable to function as desired, which could be a bad thing? If we have enough retries, won't this be resolved as long as next hit of retry against jwks server?

**Comment by incfly** [comment]
I'm trying to see if we can avoid building any additional logic in this,
and only relying on existing retries.

**Comment by haines** [comment]
>If we have enough retries, won't this be resolved as long as next hit of retry against jwks server?

What if the JWKS server has an extended outage, though? I don't think it's right that a third-party service going down would leave my sidecars stuck.

If I have multiple issuers configured, I'd want to be able to continue to accept requests from the other issuers whilst rejecting ones from the issuer whose tokens I can't validate.

**Comment by yangminzhu** [comment]
> We do not need to rely on the Envoy implementation but could just generate a different JWT filter config that rejects every request in this case.
> wondering how so?

this needs some investigation into the JWT filter API, but one quick solution is fall back to let the Envoy to fetch the remote jwks in this case.

@xulingqing could you take a look?

1. Investigate the JWT filter API to see if there is a quick way to reject the user request instead of blocking the xDS update: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/jwt_authn_filter#config-http-filters-jwt-authn
2. If that is not possible, investigate if we can just fall back to let the proxy to fetch the remote jwks, the JWT filter API already supports this and we just need to use it.

**Comment by incfly** [comment]
Fallback letting workload fetching jwks, caveat: may not work when you have networking policy restrict egress traffic. (i.e only istiod able to send http request to internet not for normal workload).

I prefer your option 1 over 2. time ordering is limited in Pilot, easier for coordinate and troubleshoot.

**Comment by md-shabbir** [comment]
I am hitting the similar issue with istio 1.5.4 with meshpolicy. if fetching the JWKS for an issuer specified in a JWT rule fails then my custom ingress gateway pod stucks in not ready state.

**Comment by vvs82** [comment]
Which of the released Istio versions (v1.6.x preferable) does have this fix?
